pkg/arvo/app/chat.hoon

```diff
@@ -340,201 +222,197 @@
-          :_  this(str.sta str)
-          %+  weld
-            [ost.bol %pull newwir [hos.affectedcir %hall] ~]~
-          %+  weld
-            (send-chat-update [[%inbox newinbox] str])
-            (send-chat-update [[%delete affectedcir] str])
-        ::  if we get a delete from another ship, delete our fake circle copy
-        ::
-        =/  deletefake/poke
-          :-  %hall-action
-              [%delete nom.fakecir ~]
-          :_  this(str.sta str)
-          %+  weld
-            [ost.bol %pull newwir [hos.affectedcir %hall] ~]~
-          %+  weld
-            [ost.bol %poke /fake [our.bol %hall] deletefake]~
-          %+  weld
-            (send-chat-update [[%inbox newinbox] str])
-            (send-chat-update [[%delete affectedcir] str])
+      :_  this(str str)
+      (send-chat-update [[%delete circ] str])
       ::
-      ::  %remove: remove a circle
-      ::
-          %remove
-        =/  str
-          %=  str.sta
-            configs   (~(del by configs.str.sta) circ)
-            messages  (~(del by messages.str.sta) circ)
-            peers     (~(del by peers.str.sta) circ)
-          ==
-        :-  (send-chat-update [[%delete circ] str])
-        this(str.sta str)
-        ::
-      ==
-      ::  end of branching on dif.sto type
     ==
-    ::  end of branching on sto type
   ==
-  ::  end of i.wir branching
 ::
-::  +lient arms
+++  handle-rumor-circle-gram
+  |=  [wir=wire sto=rumor-story:hall]
+  ^-  (quip move _this)
+  ?>  ?=(%gram -.sto)
+  =*  messages  messages.str
+  =/  circle=circle:hall  [`@p`(slav %p &2:wir) &3:wir]
+  =/  nes=(unit (list envelope:hall))  (~(get by messages) circle)
+  ?~  nes
+    [~ this]
+  =.  messages.str  (~(put by messages) circle (snoc u.nes nev.sto))
+  :_  this(str str)
+  (send-chat-update [[%message circle nev.sto] str])
 ::
+++  handle-rumor-circle-status
+  |=  sto=rumor-story:hall
+  ^-  (quip move _this)
+  ?>  ?=(%status -.sto)
+  =/  upeers=(unit (set @p))  (~(get by peers.str) cir.sto)
+  ?~  upeers
+    [~ this]
+  =/  peers=(set @p)
+    ?:  =(%remove -.dif.sto)
+      (~(del in u.upeers) who.sto)
+    (~(put in u.upeers) who.sto)
+  =.  peers.str  (~(put by peers.str) cir.sto peers)
+  :_  this(str str)
+  (send-chat-update [[%peers cir.sto peers] str])
+::
+::  +handle-rumor-circle-config-source: on source, subscribe and add to inbox
+::  on remove source, send delete and remove data from state
+::
+++  handle-rumor-circle-config-source
+  |=  sto=rumor-story:hall
+  ^-  (quip move _this)
+  ?>  ?=(%config -.sto)
+  ?>  ?=(%source -.dif.sto)
+  ?.  =(cir.sto [our.bol %inbox])
+    :: ignore when circ is not inbox
+    [~ this]
+  =*  affectedcir  cir.src.dif.sto
+  =/  newwir/wire
+    /circle/(scot %p hos.affectedcir)/[nom.affectedcir]/grams/0/config/group
+  =/  pat/path  /circle/[nom.affectedcir]/grams/0/config/group
+  ::  we've added a source to our inbox
+  ::
+  ?:  add.dif.sto
+    =.  src.inbox.str  (~(put in src.inbox.str) src.dif.sto)
```

Comment:
You can include this in the `%_  str` below. Instead of doing `inbox inbox.str`, you can do `src.inbox this-put-in`. Same below.